server/testdriver/core/src/main/java/org/infinispan/server/test/core/rollingupgrade/RollingUpgradeHandler.java

```diff
@@ -0,0 +1,229 @@
+                  handler.toDriver = handler.startNode(true, 1, nodeCount, site1Name,
+                        configuration.jgroupsProtocol(), volumeId);
+               } else {
+                  handler.toDriver.startAdditionalServer(nodeCount, volumeId);
+               }
+
+               handler.currentState = STATE.ADDED_NEW;
+
+               if (!ensureServersWorking(handler, nodeCount)) {
+                  handler.logConsumer.accept("Servers are only: " + Arrays.toString(manager.getServers()));
+                  throw new IllegalStateException("Servers did not cluster within 30 seconds, assuming error");
+               }
+            }
+
+            handler.currentState = STATE.NEW_RUNNING;
+         }
+      } catch (Throwable t) {
+         handler.currentState = STATE.ERROR;
+         configuration.exceptionHandler().accept(t, handler);
+         throw t;
+      } finally {
+         handler.cleanup();
+      }
+   }
+
+   private static boolean ensureServersWorking(RollingUpgradeHandler handler, int expectedCount) throws InterruptedException {
+      long begin = System.nanoTime();
+      while (TimeUnit.NANOSECONDS.toSeconds(System.nanoTime() - begin) < handler.configuration.serverCheckTimeSecs()) {
+         handler.logConsumer.accept("Checking to ensure cluster formed properly, expecting " + expectedCount + " servers");
+         if (handler.configuration.isValidServerState().test(handler)) {
+            return true;
+         }
+         Thread.sleep(TimeUnit.SECONDS.toMillis(5));
+      }
+      handler.logConsumer.accept("Cluster state check timed out after " + handler.configuration.serverCheckTimeSecs() +
+            " secs, check for possible issues");
+      return false;
+   }
+
+   private void cleanup() {
+      if (fromDriver != null) {
+         fromDriver.stop(configuration.fromVersion());
+      }
+      if (toDriver != null) {
+         toDriver.stop(configuration.toVersion());
+      }
+   }
+
+   public RemoteCacheManager createRemoteCacheManager() {
+      TestUser user = TestUser.ADMIN;
+
+      String hotrodURI = "hotrod://" + user.getUser() + ":" + user.getPassword() + "@" + fromDriver.getServerAddress(0).getHostAddress() + ":11222";
+      logConsumer.accept("Creating RCM with uri: " + hotrodURI);
+
+      return new RemoteCacheManager(hotrodURI);
+   }
+
+   private ContainerInfinispanServerDriver startNode(boolean toOrFrom, int nodeCount, int expectedCount, String clusterName,
+                                                            String protocol, String volumeId) {
+      ServerConfigBuilder builder = new ServerConfigBuilder("infinispan.xml", true);
+      builder.runMode(ServerRunMode.CONTAINER);
+      builder.numServers(nodeCount);
+      builder.expectedServers(expectedCount);
+      builder.clusterName(clusterName);
+      builder.property(Server.INFINISPAN_CLUSTER_STACK, protocol);
+      builder.property(TestSystemPropertyNames.INFINISPAN_TEST_SERVER_REQUIRE_JOIN_TIMEOUT, "true");
```

Comment:
Curious about this one. What would happen without the delay?

Reply:
This is needed because what this does is disable the first node created via the driver from having join timeout set to 0. This is done because normally the driver is not connecting to an existing cluster and thus it will always be the first. In our case the second driver will be connecting to the first driver cluster and we don't want to have a merge if we don't need to. Writing this though we can have this optional based on if it is the old or new version though, so I will update it.